app/services/orders/handle_fees_service.rb

```diff
@@ -16,21 +18,30 @@ def recreate_all_fees!
       # See https://github.com/rails/rails/blob/3-2-stable/activerecord/lib/active_record/locking/pessimistic.rb#L69
       # and https://www.postgresql.org/docs/current/static/sql-select.html#SQL-FOR-UPDATE-SHARE
       order.with_lock do
-        EnterpriseFee.clear_all_adjustments order
+        EnterpriseFee.clear_order_adjustments order
 
-        create_line_item_fees!
+        # To prevent issue with fee being removed when a product is not linked to the order cycle
+        # anymore, we now create or update line item fees.
+        # Previously fees were deleted and recreated, like we still do for order fees.
+        create_or_update_line_item_fees!
         create_order_fees!
       end
 
       tax_enterprise_fees! unless order.before_payment_state?
       order.update_order!
     end
 
-    def create_line_item_fees!
-      order.line_items.includes(variant: :product).each do |line_item|
-        if provided_by_order_cycle? line_item
-          calculator.create_line_item_adjustments_for line_item
+    def create_or_update_line_item_fees!
+      order.line_items.includes(:variant).each do |line_item|
+        # No fee associated with the line item so we just create them
+        if line_item.enterprise_fee_adjustments.blank?
+          create_line_item_fees!(line_item)
+          next
         end
+        create_or_update_line_item_fee!(line_item)
+
+        # delete any fees removed from the Order Cycle
+        delete_removed_fees!(line_item)
```

Comment:
We can only see what fees were removed if we still have a reference to the OC. 
If there's no reference to OC, can we skip it?
```suggestion
        delete_removed_fees!(line_item) if provided_by_order_cycle?
```

Reply:
`HandleFeeService` knows the OC via the order, so we can actually remove fees even on variant that are not provided by the OC. This is consistent with the current behaviour, so we wouldn't want to change that.

Reply:
Right, sorry for my misunderstanding